### PR TITLE
initial pass at decorrelating sub queries

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -17,14 +20,8 @@ def load_tests(loader, tests, ignore):  # pylint: disable=unused-argument
...
     """
-    from sqlglot.optimizer import qualify_columns as module1
-    from sqlglot.optimizer import qualify_tables as module2
-    from sqlglot.optimizer import quote_identities as module3
-    from sqlglot.optimizer import projection_pushdown as module4
-    from sqlglot.optimizer import scope as module5
-
-    for mod in (module1, module2, module3, module4, module5):
-        tests.addTests(doctest.DocTestSuite(mod))
+    for _, module in inspect.getmembers(sqlglot.optimizer, inspect.ismodule):
```
Comment (barakalon) [inline tests/test_optimizer.py:23]: nice

File: tests/fixtures/optimizer/rewrite_subqueries.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
+SELECT * FROM x AS x WHERE (SELECT y.a FROM y AS y WHERE x.a = y.a) = 1;
+SELECT * FROM x AS x JOIN (SELECT y.a FROM y AS y WHERE TRUE GROUP BY y.a) AS "_d_0" ON _d_0.a = x.a AND ("_d_0".a) = 1 WHERE TRUE;
```
Comment (barakalon) [inline tests/fixtures/optimizer/rewrite_subqueries.sql:2]: Should we be quoting `_d_0` in the condition too?

File: sqlglot/optimizer/rewrite_subqueries.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                eq.replace(exp.TRUE)
+                select.replace(
+                    exp.Column(
+                        this=exp.to_identifier(value.alias_or_name),
+                        table=exp.to_identifier(alias),
+                    )
+                )
+
+                condition = select.find_ancestor(
+                    exp.EQ, exp.NEQ, exp.LT, exp.LTE, exp.GT, exp.GTE
```
Comment (barakalon) [inline sqlglot/optimizer/rewrite_subqueries.py:75]: What about operators like IN?

File: tests/fixtures/optimizer/rewrite_subqueries.sql (hunk 2 above)
Comment (tobymao) [reply tests/fixtures/optimizer/rewrite_subqueries.sql:2]: i thought about that but doesn’t it happen in a later optimization step which quotes everything?

File: sqlglot/optimizer/rewrite_subqueries.py (hunk 3 above)
Comment (tobymao) [reply sqlglot/optimizer/rewrite_subqueries.py:75]: i’ll add all of these to a constant somewhere

File: sqlglot/optimizer/rewrite_subqueries.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                alias = f"_d_{next(sequence)}"
+                eq.replace(exp.TRUE)
+                select.replace(
+                    exp.Column(
+                        this=exp.to_identifier(value.alias_or_name),
+                        table=exp.to_identifier(alias),
+                    )
+                )
+
+                condition = select.find_ancestor(
```
Comment (barakalon) [inline sqlglot/optimizer/rewrite_subqueries.py:74]: I think it's possible for correlated subqueries in selects:

```sql
SELECT
    (SELECT MAX(c) FROM y WHERE y.b = x.b) AS max
FROM x
```

**Review by barakalon — APPROVED** [review]
very cool

File: tests/fixtures/optimizer/rewrite_subqueries.sql (hunk 2 above)
Comment (barakalon) [reply tests/fixtures/optimizer/rewrite_subqueries.sql:2]: Ah yeah sure,

File: sqlglot/optimizer/rewrite_subqueries.py (hunk 4 above)
Comment (tobymao) [reply sqlglot/optimizer/rewrite_subqueries.py:74]: is that optimizable one the same way? i’m sure there are many forms were missing, i’ll probably handle them later if at all

Comment (barakalon) [reply sqlglot/optimizer/rewrite_subqueries.py:74]: I think so?

```sql
SELECT 
    _d_0.max
FROM 
    x
JOIN (
    SELECT 
        b, 
        MAX(c)
    FROM 
        y
    GROUP BY b
) AS _d_0
ON _d_0.b = x.b
```

**Pull request merged**